tasks/migrate-removals.ts

```diff
@@ -0,0 +1,252 @@
+      });
+      hre.log(`Removal contract address: ${removalContract.address}`);
+      hre.log(`Signer address: ${signerAddress}`);
+      // const fireblocksSigner = removalContract.signer as FireblocksSigner;
+
+      if (!dryRun) {
+        hre.log(
+          chalk.bold.white(
+            `✨ Minting removals for ${jsonData.length} projects...`
+          )
+        );
+        hre.log(chalk.white(`🤞 Submitting transactions...`));
+      } else {
+        hre.log(
+          chalk.bold.white(
+            `DRY RUN 🌵 Minting removals for ${jsonData.length} projects...`
+          )
+        );
+      }
+      // submit all mintBatch transactions serially to avoid nonce collision!
+      const pendingTransactions = [];
+      let projectIndex = 1;
+      for (const project of jsonData) {
+        const amounts = project.amounts.map((amount: string) =>
+          ethers.utils
+            .parseUnits(divide(Number(amount), 1_000_000).toString())
+            .toString()
+        );
+
+        const removals = project.removals.map((removal: any, index) => {
+          const removalData = {
+            idVersion: Number(removal.idVersion),
+            // methodology: Number(removal.methodology), // TODO
+            // methodologyVersion: Number(removal.methodologyVersion), // TODO
+            methodology: 1,
+            methodologyVersion: 0,
+            vintage: Number(removal.vintage),
+            country: asciiStringToHexString(removal.country),
+            subdivision: asciiStringToHexString(
+              removal.subdivision.slice(0, 2)
+            ), // TODO we only want the state code here, original script was giving full substring
+            supplierAddress: '0x9A232b2f5FbBf857d153Af8B85c16CBDB4Ffb667', // TODO need real supplier address on the project
+            subIdentifier: (project.projectId % 1000) + index, // TODO get from removal
+          };
+          return removalData;
+        });
+
+        if (!dryRun) {
+          let pendingTx: ethers.ContractTransaction;
+          try {
+            pendingTx = await removalContract.mintBatch(
+              signerAddress, // mint to the consigner
+              amounts,
+              removals,
+              project.projectId,
+              // firstProject.scheduleStartTime, // TODO this can't be 0
+              // projectIndex === 1 ? 0 : 1_665_441_893, // <-- to induce a revert
+              1_665_441_893,
+              project.holdbackPercentage
+            );
+
+            pendingTransactions.push(pendingTx);
+            hre.log(
+              `(${projectIndex}/${jsonData.length}) projectId: ${
+                project.projectId
+              } txHash: ${chalk.green(pendingTx.hash)}`
+            );
+          } catch (error) {
+            hre.log(
+              `(${projectIndex}/${jsonData.length}) projectId: ${
+                project.projectId
+              } ${chalk.red('error')}`
+            );
+            pendingTransactions.push(error);
```

Comment:
should we throw instead to enforce order of on-chain removals?

Reply:
yeah I just hadn't pushed the latest... this will throw and stop when any error occurs (and also still write the file with any successful txns up to that point and the error too)